octoprint_discordremote/command.py

```diff
@@ -476,14 +478,139 @@ def unzip(self, params):
-                                 description='In case of multi-volume files, one could be missing.')
+            self.plugin.get_file_manager().remove_file('local', unzippable)
+            return None, error_embed(author=self.plugin.get_printer_name(),
+                                     title="Bad zip file.",
+                                     description='In case of multi-volume files, one could be missing.')
+
+        filelist_string = ''
+        for f in filestounpack:
+            filelist_string += (f + '\n')
+
+        return None, success_embed(author=self.plugin.get_printer_name(), title='File(s) unzipped. ', description=filelist_string)
+
+    #check if received file is eligible for unzipping
+    def judge_is_unzippable(self, filename):
+
+        autounzip = self.plugin.get_settings().get(['auto_unzip'])
+        truncated = filename[:-4]
+
+        #file is a single zip
+        if filename.endswith('.zip'):
+            #unzip
+            if autounzip:
+                return True, success_embed(author=self.plugin.get_printer_name(),
+                                           title='File Received, unzipping...',
+                                           description=filename)
+
+            #inform the user that he can unzip the file himself
+            else:
+                descr = filename + '\n'
+                descr += 'Use %sunzip %s to extract all gcode files.' % (self.plugin.get_settings().get(["prefix"]), filename)
+                return False, success_embed(author=self.plugin.get_printer_name(), title='File Received',
+                                           description=descr)
+
+        #file is part of a multi-volume zip
+        elif truncated.endswith('.zip'):
+
+            #find all available multi-volumes belonging to the file in question
+            filelist = self.get_flat_file_list()
+            available_files = []
+            available_files_sizes = []
+
+            #get all available parts of the zip
+            for f in filelist:
+
+                temp_path = f.get('path')
+
+                #check for file ending in .zip.XXX
+                if truncated.upper() in temp_path.upper():
+
+                    trunc_filename = filename[:-3] + temp_path[-3:]
+                    real_path = self.plugin.get_file_manager().path_on_disk(f.get('location'), trunc_filename)
+
+                    available_files.append(trunc_filename)
+                    available_files_sizes.append(int(os.path.getsize(real_path)))
 
-        return None, success_embed(author=self.plugin.get_printer_name(), title='File(s) unzipped. ', description=str(filestounpack))
+            smallest_filesize = available_files_sizes[0]
+            last_index = 0
+            differing_found = False
+
+            for i in range(1, len(available_files)):
+                #once we find one smaller file this must mean it's the last volume - every other file size is the same
```

Comment:
99/100 this is true, but it is possible for an exactly 8MB file, split into 2MB files to have no "smallest" file.
If this is the best/only way to do it, okay, but it definitely feels hacky at best.

Reply:
Also, I could be wrong, but i think you can theoretically have a split zip with all the files being different sizes, if you cut it up that way manually.

Reply:
Both correct, but I couldn't think of a better way of doing it.
I have checked the .zip file format specification again and apparently there is a signature in the very last file (0x06054b50) which marks the end of the central directory. I'll rework the function to check for that signature, it's unique to that file.

Reply:
Done, the code is now checking for the signature "0x06054b50" instead of going for file sizes.
The check has to be done on the last ~65kb of the volume since theoretically, users can add up to 65kb of comments and the signature is a few bytes before those.

Theoretically, if you have 100 files 1kb each and the last 30 files are only comments, the code will assume that you have only 70 volumes since it found the signature there - unzipping will still work with those first 70 files, though.